src/runtime.rs

```diff
@@ -153,7 +149,453 @@ pub struct FsSearchBody {
+                let mut index_to_data_hash_map = HashMap::new();
+                for info in shared_load_data_infos.iter() {
+                    let hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                    let idx = file_info_indices[usize::from(info.file_info_indice_index)].file_info_index;
+                    let idx = file_infos[usize::from(idx)].file_path_index;
+                    index_to_data_hash_map.insert(idx, hash);
+                }
+                // we do this to be able to go from shared data hash -> mass load group hash
+                // this is important so that the file path indices in our new MassLoadData point
+                // to the original filepaths
+                let mut data_to_group_hash_map = HashMap::new();
+                let mut group_to_index_hash_map = HashMap::new();
+                for (offset, info) in mass_load_group_infos.iter().enumerate() {
+                    let idx = file_info_indices[usize::from(info.file_info_indice_index)].file_info_index;
+                    let idx = file_infos[usize::from(idx)].file_path_index;
+                    if let Some(data_hash) = index_to_data_hash_map.get(&idx) {
+                        let group_hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                        if *data_hash == group_hash {
+                            return Err(LookupError::Missing); // TODO: Change this to a better error code or smthn
+                        }
+                        data_to_group_hash_map.insert(*data_hash, group_hash);
+                        group_to_index_hash_map.insert(group_hash, offset);
+                    }
+                }
+                drop(index_to_data_hash_map);
+
+                // since we also have to include our redirection chain, we have to make sure we get contigious FileInfos
+                // for our new MassLoadData
+                let current_mld_offset = new_infos.len();
+                new_infos.resize_with(current_mld_offset + shared_load_data_infos.len(), || default_info);
+                for (offset, info) in shared_load_data_infos.iter().enumerate() {
+                    // get the shared hash
+                    let source_info_hash = file_paths[usize::from(info.file_path_index)].path.hash40();
+                    let new_info_hash = data_to_group_hash_map.get(&source_info_hash).expect(&format!("Could not find new hash for source {:#x?}", source_info_hash));
+                    // extract the MassLoadGroup's child info for this
+                    let child_info_offset = group_to_index_hash_map.get(&new_info_hash).expect(&format!("Could not find info index for new hash {:#x?}", new_info_hash));
+                    let child_info = &mut mass_load_group_infos[*child_info_offset];
+                    // get the data this pointed to
+                    let idx_ = child_info.file_info_indice_index;
```

Comment:
not sure why you're using `idx_`? either shadow `idx` if you don't need both at once or use a more specific name.